frontend/src/components/resourceMap/useQueryParamsState.tsx

```diff
@@ -0,0 +1,57 @@
+import { Dispatch, SetStateAction, useEffect, useState } from 'react';
+import { useLocation } from 'react-router';
+
+type UseQueryParamsStateReturnType<T> = [T, Dispatch<SetStateAction<T>>];
+export function useQueryParamsState<T extends string | undefined>(
+  param: string,
+  initialState: T
+): UseQueryParamsStateReturnType<T> {
+  const location = useLocation();
+
+  // State for managing the value derived from the query parameter
+  const [value, setValue] = useState<T>(() => {
+    if (typeof window === 'undefined') return initialState;
+
+    // Parse query parameter value from the URL
+    const { search } = window.location;
+    const searchParams = new URLSearchParams(search);
+    const paramValue = searchParams.get(param);
+
+    return paramValue !== null ? (decodeURIComponent(paramValue) as T) : initialState;
+  });
+
+  useEffect(() => {
```

Comment:
A named function here instead of an anonymous one will document what this function does better.

Reply:
we never use named functions inside an effect. for consistency keeping it as an arrow function but adding a comment above it